src/main/java/net/helix/pendulum/controllers/RoundViewModel.java

```diff
@@ -242,72 +250,76 @@ public static int getRoundIndex(TransactionViewModel milestoneTransaction) {
         // idx = n: milestone merkle root in trunk
         if (transaction.getCurrentIndex() == transaction.lastIndex()) {
-            // add previous milestones to non analyzed transactions
-            RoundViewModel prevMilestone = RoundViewModel.get(tangle, round-1);
-            if (prevMilestone == null) {
-                if (transaction.getBranchTransactionHash().equals(Hash.NULL_HASH)) {
-                    trunk.add(Hash.NULL_HASH);
-                }
-            } else {
-                Set<Hash> prevMilestones = prevMilestone.getHashes();
-                List<List<Hash>> merkleTree = Merkle.buildMerkleTree(new ArrayList<>(prevMilestones));
-                if (transaction.getTrunkTransactionHash().equals(merkleTree.get(merkleTree.size() - 1).get(0))) {
-                    if (prevMilestones.isEmpty()) {
-                        trunk.add(Hash.NULL_HASH);
-                    } else {
-                        trunk.addAll(prevMilestones);
-                    }
-                }
-            }
-        }
-        else {
+            trunk.addAll(splitMilestonesMerkleRoot(tangle, transaction.getTrunkTransactionHash(), milestoneTx));
+        } else {
             // idx = 0 - (n-1): merkle root in branch, trunk is normal tx hash
             trunk.add(transaction.getTrunkTransactionHash());
         }
         return trunk;
     }
 
-    public static Set<Hash> getMilestoneBranch(Tangle tangle, TransactionViewModel transaction, TransactionViewModel milestoneTx, int security) throws Exception{
+
+    public static Set<Hash> getMilestoneBranch(Tangle tangle, TransactionViewModel transaction, TransactionViewModel milestoneTx, int security) throws Exception {
         Set<Hash> branch = new HashSet<>();
-        int round = RoundViewModel.getRoundIndex(milestoneTx);
         // idx = n: milestone merkle root in trunk and tips merkle root in branch
         if (transaction.getCurrentIndex() == transaction.lastIndex()) {
-            // tips merkle root
-            Set<Hash> confirmedTips = getTipSet(tangle, milestoneTx.getHash(), security);
-            List<List<Hash>> merkleTree = Merkle.buildMerkleTree(new ArrayList<>(confirmedTips));
-            if (transaction.getBranchTransactionHash().equals(merkleTree.get(merkleTree.size()-1).get(0))) {
-                if (confirmedTips.isEmpty()){
-                    branch.add(Hash.NULL_HASH);
+            branch.addAll(splitTipsMerkleRoot(tangle, transaction, milestoneTx, security));
+        } else {
+            branch.addAll(splitMilestonesMerkleRoot(tangle, transaction.getBranchTransactionHash(), milestoneTx));
+        }
+
+        if (log.isTraceEnabled()) {
+            log.trace("Milestone branch: {}", PendulumUtils.logHashList(branch, 8));
+        }
+        return branch;
+    }
+
+
+    private static Set<Hash> splitMilestonesMerkleRoot(Tangle tangle, Hash expectedMilestoneMerkleRoot, TransactionViewModel milestoneTx) throws Exception {
```

Comment:
In view of our last discussion, does it make sense to simply pack milestone hashes into the bundle similar to the tips?

Reply:
It could be a solution, but what is the purpose of virtual transactions then?

Reply:
We will probably not use them at all, at least for now